README.md
# craft ai energy kit #

> The [craft ai] integration kit for energy prediction in JavaScript.
<!-- -->
> :construction: This project is still under active development.

[![Version](https://img.shields.io/npm/v/craft-ai-kit-energy.svg?style=flat-square)](https://npmjs.org/package/craft-ai-kit-energy)
[![Build](https://img.shields.io/travis/craft-ai/craft-ai-kit-energy/master.svg?style=flat-square)](https://travis-ci.org/craft-ai/craft-ai-kit-energy)
[![Code coverage](https://img.shields.io/coveralls/github/craft-ai/craft-ai-kit-energy.svg?style=flat-square)](https://coveralls.io/github/craft-ai/craft-ai-kit-energy?branch=master)
[![License](https://img.shields.io/badge/license-BSD--3--Clause-42358A.svg?style=flat-square)](LICENSE)
[![Dependencies](https://img.shields.io/david/craft-ai/craft-ai-kit-energy.svg?style=flat-square)](https://david-dm.org/craft-ai/craft-ai-kit-energy)
[![Dev Dependencies](https://img.shields.io/david/dev/craft-ai/craft-ai-kit-energy.svg?style=flat-square)](https://david-dm.org/craft-ai/craft-ai-kit-energy#info=devDependencies)

## Overview ##

The [craft ai] energy kit is a plug-and-play JavaScript library allowing to leverage the [craft ai] explainable artificial intelligence for predicting energy consumption behaviors.

For each of the device that you want to modelize, you will define an _endpoint_.
This endpoint will be identified by a unique _id_ string. You might also provide some details as to the location.

### Workflow ###

1. Initialize: the endpoint accepts a historical record of power or energy as input; based on this record, it will generate a predictive model.

2. Evaluate: make sure that the model is able to capture the behavior of your signal.

3. Exploit: a model allows you to
    - detect anomalies in the past;
    - predict the behavior of your device in the future.

4. Update: predictive models are dynamic and it is possible to update them with new data at any time; Feeding them newly recorded data allows to keep them relevant.

## Usage ##

###

This integration kit has been tested on every active and maintained [`LTS` releases of Node.js](https://github.com/nodejs/Release#release-schedule) (i.e. it should work with any version greater or equal to `v6.9.0`).

#### Add the energy kit to your project ####

```console
$ npm i craft-ai-kit-energy
```

#### Initialize an instance of the kit ####

> All the functions exposed by the kit return native [**Promises**](http://www.datchley.name/es6-promises/).

```js
const EnergyKit = require('craft-ai-kit-energy');

EnergyKit
  .initialize({
    /**
     * __Required__
     *
     * Specify the craft ai token with write access to a project. The kit will
     * use this project to host its agents. You can also specify it through the
     * environment variables `CRAFT_AI_TOKEN`.
     */
    token: '{the-craft-ai-token}',
    /**
     * __Recommended in a production environnement.__
     *
     * Define a string used to derive and anonymise the unique identifiers of
     * an endpoint when creating the related craft ai agents. If not specified,
     * each agent will use the same identifier as the endpoint's identifier.
     */
    secret: '{a-secret-string}',
    /**
     * __Optional__
     *
     * Specify external data providers to enrich your records.
     * Preset providers are 'PublicHolidayProvider', 'SchoolHolidaysProvider' (France only)
     * and 'WeatherProvider' using the Dark Sky API.
     */
    providers: [
      {
        provider: PublicHolidayProvider,
        options: { country: 'fr' }
      },
      {
        provider: SchoolHolidaysProvider,
        options: { country: 'fr' }
      },
      {
        provider: WeatherProvider,
        options: {
          /**
           * __Required__
           *
           * Specify the access token to the weather provider.
           */
          token: '{dark-sky-token}',
          /**
           * __Optional__
           *
           * List of properties to retrieve from the weather provider.
           * Default to ['temperatureLow', 'temperatureHigh'].
           */
          properties: [
            '{a-first-property-to-retrieve}',
            '{a-second-property-to-retrieve}'
          ],
          /**
           * __Optional__
           *
           * Enable the cache mechanism.
           */
          cache: {
            load: () => require('{weather-cache-path}'),
            save: (cache) => {
              fs.writeFileSync(
                '{weather-cache-path}',
                JSON.stringify(cache, null, '  ')
              );
            }
          }
        }
      }
    ]
  })
  .then((kit) => /* Do something with the kit... */);
```

#### Load an endpoint ####

One instance of the kit can manage several electrical devices, they are called _endpoints_. They need at least a unique identifier to be loaded.

```js
kit
  .loadEndpoint(
  /**
   * __Required__
   *
   * Endpoint configuration.
   */
    {
    /**
     * __Required__
     *
     * Define uniquely the endpoint.
     */
    id: '{a-unique-id}',
    /**
     * __Optional__
     *
     * Add details as to what the endoint represents.
     * Some metadata properties might be required by the providers.
     */
    metadata: {
      region: '{2-digit-department-code}', // Used by the default packaged School Holidays provider (France only)
      latitude: '{latitude}',  // Used by the default Weather provider
      longitude: '{longitude}'   // Used by the default Weather provider
    }
    energy: __
    agent_id: '{craft-ai-agent-id}', // Id of the craft ai agent, default = id
    learning: {
      maxTreeDepth: int, // Max depth of the tree, default = 6
      maxRecords: int, // Max operations in the tree, default = 50000
      maxRecordAge: int, // Learning period, default = 31536000 (1 year)
      properties: {
        feature1: {type: 'continuous'}, // Feature of type continuous
	feature2: {type: 'enum'}, // Feature of type enumeration
	...
      }
    }
  },
  /**
   * __Optional__
   *
   * Specify if the endpoint must be reset.
   * true => delete the associated agent if it exists and recreate it.
   * false => retrieve the associated agent if it exists, create it otherwise.
   * Default to false.
   */
  false)
  .then((endpoint) => /* Do something with the endpoint... */)
```

#### Update the endpoint ###

Before using the endpoint, you will want to update it with energetic consumption data.
The endpoint takes as an input a _record_, which consists at least of a dated energetic information.
This energetic data can be one of the two following forms:

- `load` (=instant power);
- `energy` (transformed energy since last index)

```js
endpoint
  .update([
    { date: new Date(2018, 0, 1), load: 740 },
    { date: new Date(2018, 0, 2), load: 415 },
    { date: new Date(2018, 0, 3), load: 609 },
    /* ... */
    { date: new Date(), load: 286 },
  ])
  .then((endpoint) => /* Do something with the updated endpoint... */)
```

#### Harness the endpoint's predictive model ####

Once your endpoint instance is up-to-date, you can make use of its underlying predictive model to retrieve for example a list of consumption anomalies.

> For a complete list of available methods, refer to [the API reference](#api-reference).

```js
const today = new Date()
const lastMonth = new Date(today - 30 * 24 * 3600 * 1000)

endpoint
  .computeAnomalies(
  /**
   * __Required__
   *
   * Date range for which to compute anomalies.
   * From {date} to {date}.
   */
    { from: lastMonth, to: today },
  )
  .then((anomalies) => console.log(anomalies))
```

#### Close the kit ####

When you're done using the kit, close it:

```js
kit
  .close()
  .then(() => /* Do something after the kit has been closed... */)
```

> This will also close the connection with any data providers specified at initialization.

#### Adding additional context information ####

In most cases, the endpoint's consumption doesn't only depends on time. You might want to boost the underlying predictive model with weather information or holidays.

> This feature is not documented yet. Contact us or jump into the code, for more information.

### API reference ###

#### `endpoint.computeAnomalies({ from: Date, to: Date })` ####

Detect the anomalies for mean electrical load based on the stored history of the endpoint in a given range of time.

#### `endpoint.computePredictions(Array<State>)` ####

> A state consists in records object without the electrical data.

Retrieve the predictions of mean electrical load for a set of states.

#### `endpoint.computeReport({ from: Date, to: Date })` ####

Retrieve some statistics for mean electrical load based on the stored history of the endpoint in a given range of time.

#### `endpoint.destroy()` ####

Delete the endpoint, all its stored records from the related craft ai agent as well as the agent itself.

#### `endpoint.retrievePredictiveModel(modelDate?: Date)` ####

Retrieve the craft ai predictive model at a given date. By default, the model based on the last stored endpoint's records will be downloaded.

#### `endpoint.retrieveRecords(from: Date, to: Date)` ####

Retrieve the stored endpoint's records from the related craft ai agent.

#### `endpoint.update(Array<Record>)` ####

Update the endpoint by adding the new records to the related craft ai agent. Older records than the one already stored on the agent are simply ignored.

#### Environment variables ####

The kit uses the following environment variables for secrets and global configuration:

- `CRAFT_AI_TOKEN` the [craft ai] token with write access to a project. The kit will use this project to host its agents.
- `DEBUG` (*optional*) a prefix to toggle the debug output for different parts of the kit. Set it to `craft-ai:*` to display the logging information related to [craft ai]. For details, refer to [the `debug` package](https://github.com/visionmedia/debug).

> ℹ To manage the environment variables of your project, you can use the [`dotenv` project](https://github.com/motdotla/dotenv).

## Examples ##

### Data preparation ###

> The following steps are optional as the prepared datasets are included on the repository.

#### [AMPds2: The Almanac of Minutely Power dataset (Version 2)] ####

The script [`./examples/retrieve_ampds2.js`](examples/retrieve_ampds2.js) downloads and prepares the dataset to [`./examples/data/ampds2.csv`](examples/data/ampds2.csv) in order to use it in other examples. On a good internet connection it should last ~10 minutes.

```console
$ npm run examples:retrieve_ampds2
```

#### [UCI Individual Household Electric Power Consumption Data Set] ####

The script [`./examples/retrieve_uci_household_power_consumption.js`](examples/retrieve_uci_household_power_consumption.js) downloads and prepares the dataset to [`./examples/data/uci_household_power_consumption.csv`](examples/data/uci_household_power_consumption.csv) in order to use it in other examples. On a good internet connection it should last ~10 minutes.

```console
$ npm run examples:retrieve_uci_household_power_consumption
```

### Single endpoint example ###

The script [`./examples/single_endpoint.js`](examples/single_endpoint.js) showcases most of the features of the kit using the aformentioned [UCI Individual Household Electric Power Consumption Data Set]. It builds an endpoint from the dataset, enrich its data with weather and holidays information and uses the kit to compute anomalies, report and predictions.

```console
$ npm run examples:single_endpoint
```

## Attribution ##

[![Powered By Dark Sky](./poweredby_dark_sky.png)](https://darksky.net/poweredby/)

> Dua, D. and Karra Taniskidou, E. (2017). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]([http://archive.ics.uci.edu/ml). Irvine, CA: University of California, School of Information and Computer Science.
<!-- -->
> Hébrail, Georges and Bérard, Alice, 2012 "Individual household electric power consumption Data Set", [https://archive.ics.uci.edu/ml/datasets/individual+household+electric+power+consumption](https://archive.ics.uci.edu/ml/datasets/individual+household+electric+power+consumption)
<!-- -->
> Makonin, Stephen, 2016, "AMPds2: The Almanac of Minutely Power dataset (Version 2)", [https://doi.org/10.7910/DVN/FIE0S4](https://doi.org/10.7910/DVN/FIE0S4), Harvard Dataverse, V1, UNF:6:0uqZaBkSWdyv27JqTHFWPg==\[fileUNF\]

[craft ai]: http://www.craft.ai
[AMPds2: The Almanac of Minutely Power dataset (Version 2)]: https://doi.org/10.7910/DVN/FIE0S4
[UCI Individual Household Electric Power Consumption Data Set]: https://archive.ics.uci.edu/ml/datasets/individual+household+electric+power+consumption


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.6...HEAD) ##

### Fixed

- Minor changelog fix

## [0.4.6](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.5...v0.4.6) - 2020-10-30 ##

### Changed

- Predictions on models which are not based on any data point return `null` results instead of throwing an error.

## [0.4.5](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.4...v0.4.5) - 2020-09-04 ##

### Added

- Handle concurrence when loading endpoint

### Fixed

- Fix the version of the craft ai package

## [0.4.4](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.3...v0.4.4) - 2020-08-25 ##

### Added

- Add js doc

### Fixed

- Update versions of packages

## [0.4.3](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.2...v0.4.3) - 2019-11-14 ##

### Fixed

- Fix lint issues
- Fix tests with custom configuration parameter

## [0.4.2](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.1...v0.4.2) - 2019-11-14 ##

### Fixed

- Enable `roundDate` to have period equal to NaN

## [0.4.1](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.4.0...v0.4.1) - 2019-10-23 ##

### Fixed

- Fixing the deploy version of node

## [0.4.0](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.3.2...v0.4.0) - 2019-10-22 ##

### Added

- Introducing `advancedConfiguration` in the kit to add options to the craft ai configuration

### Changed

- Updating dependencies, the package [**craft ai**](https://www.npmjs.com/package/craft-ai) is now in version 1.17.1
- Discontinuing the support for Node 6.9

## [0.3.2](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.3.1...v0.3.2) - 2019-05-30 ##

### Fixed

- Fixing the usage of `dotenv` in the examples code.
- Sample retrieval code for AMPds2 now uses the right unit.

### Changed

- Updating dependencies, especially [**craft ai**](https://www.npmjs.com/package/craft-ai) now using v2 of the decision trees.

## [0.3.1](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.3.0...v0.3.1) - 2019-05-02

### Fixed

- Parsing numeric embedded properties when provided as a string.
- Fixed offset zones like `UTC+8` or `UTC-03:30` were considered invalid zones.

### Added

- Caching `zone` information to speed up records parsing and prevent high CPU loads.
- Introducing the `endpoint.evaluate()` method, to evaluate an endpoint with a rolling prediction task.
- Introducing an integration test that uses the new evaluate method to keep track of the kit's
  performances on two prediction tasks.

### Changed

- Energy data is now defined with an origin as a datetime string representing a **local** time and a period as a number of seconds.

## [0.3.0](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.2.2...v0.3.0) - 2018-11-20

### Fixed

- Unable to parse dates with negative time offsets.
- Unable to compute predictions on states outside the local timezone.
- Unbale to compute predictions from retrieved records without manually setting the timezone

### Added

- Retrieving the timezone of a record directly from its `timezone` feature.
- Introducing a `zone` option on the kit and endpoint's objects.

## [0.2.2](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.2.1...v0.2.2) - 2018-10-08

### Fixed

- Caching issue with the `WeatherProvider` while extending records on a hourly basis with offseted times.
- Shifted Easter based holidays when given a non-UTC date to the `PublicHolidayProvier`.

## [0.2.1](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.2.0...v0.2.1) - 2018-09-28

### Fixed

- Unable to use the French `SchoolHolidaysProvider`.

## [0.2.0](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.1.0...v0.2.0) - 2018-09-27

### BREAKING CHANGES

The signature of the method `endpoint.computePredictions()` has been changed to include an `options` argument.

### Added

- Introducing a CSV import helper with streaming capabilities and custom parsing options.
- Supporting a path to a CSV file as the first argument of some methods of an endpoint's instance.
- Introducing a `WeatherProvider` [powered by _Dark Sky API_](https://darksky.net/poweredby/).
- Exposing an option to control the refresh rate of the providers.
- Checking the optional field `metadata` of the endpoints.
- Supporting additional embedded context properties.
- Handling records with consumption as energy and accumulated energy.
- Exposing options to load and save the internal cache of the `WeatherProvider`.
- Introducing an **examples** directory containing, for now, one simple example showcasing the kit using [UCI's Individual Household Electric Power Consumption Data Set](https://archive.ics.uci.edu/ml/datasets/individual+household+electric+power+consumption).

### Changed

- Clearing caches when closing a provider.
- Passing the whole record to the provider for the generation of record extensions.
- Rename the features used by the public and school holiday providers to be able to use both.

### Fixed

- Unable to compute predictions with states (not containing `load` property).
- Unable to compute predictions while using providers.
- Exception raised by the `WeatherProvider` when issuing bad request to _Dark Sky API_.

## [0.1.0](https://github.com/craft-ai/craft-ai-kit-energy/compare/v0.0.7...v0.1.0) - 2018-07-04

### BREAKING CHANGES

In the aim of publishing the first version of the energy kit, we rewrote its API from the ground up:

- introducing the notion of _endpoint_, which interfaces with a craft ai agent by uploading the consumption history of an electrical endpoint and exposing a few methods to make the most of the underlying predictive models.

src/kit.js
const craftaiErrors = require('craft-ai/lib/errors');
const debug = require('debug');
const uuid = require('uuid/v5');

const Constants = require('./constants');
const Endpoint = require('./endpoint');
const Provider = require('./provider');
const Utils = require('./utils');

const DEBUG_PREFIX = Constants.DEBUG_PREFIX;
const LOAD = Constants.LOAD_FEATURE;
const TIMEZONE = Constants.TIMEZONE_FEATURE;

/**
 * @typedef Energy
 * @type {object}
 * @property {Number} period positive number (in seconds)
 * @property {Number} hours positive numver
 * @property {String} origin date which is the origin.
 * @property {Provider} providers
 */

/**
 * @typedef Endpoint
 * @type {object}
 * @property {String} id name of the endpoint
 * @property {Energy} [energy] [OPTIONAL] energy configuration
 * @property {Object} [metadata] [OPTIONAL]
 * @property {String} [agentId] [OPTIONAL] identifiant of the craft ai agent, Default is id.
 * @property {String} [agent] [OPTIONAL]
 * @property {Object} [learning] [OPTIONAL] as described in generateAgentConfiguration
 * @property {Object} [kit] [OPTIONAL]
 * @property {Object} [feature] [OPTIONAL]
 * @property {Object} generated
 */

/**
 * Get the craft ai agent endpoint
 *
 * @param {Endpoint} definition
 * @param {boolean} [resetAgent=false] [OPTIONAL], if true the agent is deleted and (re-)created
 *
 * @returns {Object} An object endpoint
 */
async function loadEndpoint(definition, resetAgent = false) {
  if (definition === null || typeof definition !== 'object') {
    throw new TypeError(`The endpoint's definition must be an "object". Received "${definition === null ? 'null' : typeof definition}".`);
  }

  const id = definition.id;
  const log = debug(`${DEBUG_PREFIX}:endpoint:${id}`);

  log('loading');

  if (typeof id !== 'string') {
    throw new TypeError(`The "id" property of the endpoint's definition must be a "string". Received "${id === null ? 'null' : typeof id}".`);
  }

  const metadata = definition.metadata;

  if (metadata !== undefined) {
    if (metadata === null || typeof metadata !== 'object') {
      throw new TypeError(`The "metadata" property of the endpoint's definition must be an "object". Received "${metadata === null ? 'null' : typeof metadata}"`);
    }

    const zone = metadata.zone;

    if (zone === undefined) {
      metadata.zone = this.configuration.zone;
    }
    else {
      if (Utils.isNotString(zone)) {
        throw new TypeError(`The "zone" property of the endpoint's configuration must be a "string". Received "${typeof zone}".`);
      }
      if (!Utils.checkZone(zone)) {
        throw new RangeError('The "zone" property of the endpoint\'s configuration must be a valid IANA zone or a fixed-offset name.');
      }
    }
  }

  const energy = parseEnergyConfiguration(definition.energy);
  const namespace = this.configuration.namespace;
  const agentId = definition.agentId || (namespace ? uuid(id, namespace) : id);

  return generateAgentConfiguration(log, this.configuration.providers, definition.learning)
    .then((agentConfiguration) => retrieveAgent(log, this.client, agentId, agentConfiguration, resetAgent))
    .then((agent) => {
      const context = agent.configuration.context;
      const contextKeys = Object.keys(context);
      const features = contextKeys.filter((feature) => !context[feature].is_generated);
      const generated = contextKeys.filter((feature) => context[feature].is_generated);

      log('loaded');

      return Object.create(Endpoint, {
        agent: { value: agent, configurable: true },
        debug: { value: log },
        energy: { value: energy },
        features: { value: features },
        generated: { value: generated },
        kit: { value: this },
        agentId: { value: agentId, enumerable: true },
        id: { value: id, enumerable: true },
        metadata: { value: metadata || { zone: this.configuration.zone }, enumerable: true }
      });
    });
}

/**
 * Close the kit and all of its providers
 */
async function close() {
  return Provider
    .close(this.configuration.providers)
    .then(() => this.debug('closed'));
}

/**
 * Generate a craft ai agent configuration
 *
 * @param {String} log
 * @param {Array<Provider>} providers Array of providers that have as function extendConfiguration
 * @param {Object} [learning={}] [OPTIONAL] Object that contains parameters of the configuration. Can contain:
 *  - maxTreeDepth: positive integer, define the max depth of the decision tree. Default is 6.
 *  - maxRecords: positive integer, maximum number of events on which a single decision tree can be based. Default is 50000.
 *  - maxRecordAge: positive integer, the maximum amount of time, in seconds, that matters for an agent. Default is one year.
 *  - properties: object, context properties to add to {time: { type: 'time_of_day' }, day: { type: 'day_of_week' },
 *    month: { type: 'month_of_year' }, timezone: { type: 'timezone' }, load: { type: 'continuous' }}
 *  - advancedConfiguration: Object, advanced configuration options. Default is {}
 * @returns {Object} craft ai agent configuration
 */
async function generateAgentConfiguration(log, providers, learning = {}) {
  log('generating the agent\'s configuration');

  if (learning === null || typeof learning !== 'object') {
    throw new TypeError(`The "learning" property of the endpoint's definition must be an "object". Received "${ learning === null ? 'null' : typeof learning }".`);
  }

  const maxTreeDepth = learning.maxTreeDepth;
  if (maxTreeDepth !== undefined && typeof maxTreeDepth !== 'number') {
    throw new TypeError(`The "maxTreeDepth" property of the endpoint's learning definition must be a "number". Received "${maxTreeDepth === null ? 'null' : typeof maxTreeDepth}".`);
  }

  const maxRecords = learning.maxRecords;

  if (maxRecords !== undefined && typeof maxRecords !== 'number') {
    throw new TypeError(`The "maxRecords" property of the endpoint's learning definition must be a "number". Received "${maxRecords === null ? 'null' : typeof maxRecords}".`);
  }

  const maxRecordAge = learning.maxRecordAge;

  if (maxRecordAge !== undefined && typeof maxRecordAge !== 'number') {
    throw new TypeError(`The "maxRecordAge" property of the endpoint's learning definition must be a "number". Received "${maxRecordAge === null ? 'null' : typeof maxRecordAge}".`);
  }

  const properties = learning.properties;

  if (properties !== undefined && (properties === null || typeof properties !== 'object')) {
    throw new TypeError(`The "properties" property of the endpoint's learning definition must be an "object". Received "${properties === null ? 'null' : typeof properties}".`);
  }

  const advancedConfiguration = learning.advancedConfiguration;

  if (advancedConfiguration !== undefined && (advancedConfiguration === null || typeof advancedConfiguration !== 'object')) {
    throw new TypeError(`The "options" property of the endpoint's learning definition must be an "object". Received "${advancedConfiguration === null ? 'null' : typeof advancedConfiguration}".`);
  }

  return Provider.extendConfiguration(providers, {
    time: { type: 'time_of_day' },
    day: { type: 'day_of_week' },
    month: { type: 'month_of_year' },
    ...properties,
    [TIMEZONE]: { type: 'timezone' },
    [LOAD]: { type: 'continuous' }
  })
    .then((context) => {
      return {
        context,
        output: ['load'],
        operations_as_events: true,
        tree_max_depth: maxTreeDepth || 6,
        tree_max_operations: maxRecords || 50000,
        learning_period: maxRecordAge || 365 * 24 * 60 * 60,
        ...advancedConfiguration
      };
    });
}

/**
 * Get the craft ai agent
 *
 * @param {String} log
 * @param {Object} client craft ai client
 * @param {String} agentId id given to the agent to get
 * @param {Object} agentConfiguration craft ai agent configuration, it is needed if the agent
 *  is reseted or doesn't exist and needs to be created
 * @param {boolean} [resetAgent] [OPTIONAL] if true the agent is deleted and created
 *
 * @returns {Object} craft ai agent configuration
 */
async function retrieveAgent(log, client, agentId, agentConfiguration, resetAgent) {
  if (resetAgent) {
    /* istanbul ignore if */
    if (process.env.NODE_ENV === 'production') {
      console.warn('WARNING: The endpoint\'s agent is being forced to be resetted.');
    }

    return client
      .deleteAgent(agentId)
      .then(() => createAgent(log, client, agentId, agentConfiguration));
  }

  log('retrieving the agent "%s"', agentId);

  return client
    .getAgent(agentId)
    // TODO: Check the configuration of the retrieved agent
    .then((agent) => {
      log('the agent has been retrieved');

      return agent;
    })
    .catch((error) => {
      /* istanbul ignore else */
      if (error instanceof craftaiErrors.CraftAiBadRequestError && error.message.includes('[NotFound]')) {
        log('the agent does not exist');

        return createAgent(log, client, agentId, agentConfiguration);
      }

      /* istanbul ignore next */
      // TODO: proper error handling
      throw error;
    });
}

/**
 * Create a craft ai agent
 *
 * @param {String} log
 * @param {Object} client craft ai client
 * @param {String} agentId id to give to the agent
 * @param {Object} agentConfiguration craft ai agent configuration
 *
 * @returns {Object} craft ai agent configuration
 */
async function createAgent(log, client, agentId, agentConfiguration) {
  log('creating the agent');

  return client
    .createAgent(agentConfiguration, agentId)
    .then(() => {
      log('the agent has been created');
      return client.getAgent(agentId);
    })
    .catch(/* istanbul ignore next */(error) => {
      // Agent has been already created during our creation so we just get the info from it
      if (error.message.includes('because one already exists.')) {
        return client.getAgent(agentId);
      }
      throw error;
    });
}

/**
 * Parse the energy configuration
 *
 * @param {Energy} energy
 *
 * @returns {{ period: Number, hours: Number, origin: String }} Contains the parsed energy configuration
 */
function parseEnergyConfiguration(energy) {
  const parsedEnergy = {};

  if (energy === undefined) {
    return energy;
  }
  if (energy === null || typeof energy !== 'object') {
    throw new TypeError(`The "energy" property of the endpoint's definition must be an "object". Received "${energy === null ? 'null' : typeof energy}".`);
  }

  if (energy.period !== undefined) {
    if (typeof energy.period !== 'number') {
      throw new TypeError(`The "period" property of the endpoint's energy definition must be an "number". Received "${typeof energy.period}".`);
    }

    if (energy.period <= 0) {
      throw new RangeError(`The "period" property of the endpoint's energy definition must represent a strictly positive duration. Received "${energy.period}".`);
    }

    parsedEnergy.period = energy.period * 1000;
    parsedEnergy.hours = energy.period / 3600;
  }

  if (energy.origin !== undefined) {
    const origin = Utils.parseDate(energy.origin);

    if (!origin.isValid) {
      throw new RangeError(`The "origin" property of the endpoint's energy definition must be a valid date definition. Received "${JSON.stringify(energy.origin)}". Reason: ${origin.invalidReason}.`);
    }

    if (energy.period === undefined) {
      throw new Error('The "origin" property of the endpoint\'s energy definition cannot be defined without a "period" property.');
    }

    parsedEnergy.origin = origin;
  }

  return parsedEnergy;
}

module.exports = {
  loadEndpoint,
  close
};
